### ISSUE TITLE
Unable to train

### ISSUE THREAD

**Issue by mustafaakinyilmaz** [InterDigitalInc/CompressAI#4]
First of all thank you for your amazing work. I have a question about training. I am trying to train MeanScaleHyperprior network with (192, 192) channels using vimeo90k set. I randomly crop 256*256 patches, use 16 as batch size and set lamda=0.0018 (quality 1). However I am not able to properly train the network so that the results match with the plots on https://interdigitalinc.github.io/CompressAI/zoo.html. There is one small difference on reading images. I use imageio.imread which returns 0-255 range, then I normalize by dividing 255. I observe no significant change after 500k iterations and stuck at 26.86 dB, 0.30 bpp on Kodak dataset. Do you have any guess what may be the reason? Thanks

**Comment by jbegaint** [comment]
Hi! yes this does not look right. are you using the example training script? what's your learning rate?
If you can provide you training code this could be helpful also!
(I don't think this should be related to using imageio)

**Comment by mustafaakinyilmaz** [comment]
I am using the following code for training. 


```
# coding: utf-8

# In[1]:


import torch
import torch.nn as nn
from torch.autograd import Variable
import numpy as np
import glob
import os
import itertools
import warnings
warnings.filterwarnings('ignore')
import imageio
from natsort import natsorted
from skimage.measure import compare_ssim,compare_psnr,compare_mse
import logging
import time
import random
import torch.nn.functional as F

from compressai.models import MeanScaleHyperprior
from rd import RateDistortionLoss

device = torch.device("cuda:0")

logging.basicConfig(filename="compression.log",level=logging.INFO)


def normalize(tensor):
    return tensor/255.
    
def float_to_uint8(image):
    clip = np.clip(image*255,0,255)
    im_uint8 = np.round(clip).astype(np.uint8).transpose(1,2,0)
    return im_uint8

def MSE(gt,pred):
    mse = np.mean((gt-pred)**2)
    return mse

def PSNR(mse,data_range):
    psnr = 10*np.log10((data_range**2)/mse)
    return psnr

def calculate_distortion_loss(out,real):
    distortion_loss = torch.mean((out-real)**2)
    return distortion_loss

def calculate_rate_loss(likelihood):
    return -torch.sum(torch.log2(likelihood))



def train_image_loader(path="/home/ml/Akin/vimeo_septuplet/sequences/"):
    images = []
    fold = natsorted(glob.glob(path+"*"))
    for f in fold:
        mfold = natsorted(glob.glob(f+"/*"))
        for m in mfold:
            imgs = natsorted(glob.glob(m+"/*.png"))
            for i in imgs:
                images.append(i)
    return images


def prepare_train_data(image_batch):
    
    X_train = []
    
    patch_size = 256
    
    for ims in image_batch:
        
        img_orig = imageio.imread(ims)
        
        x = random.randint(0, img_orig.shape[1] - patch_size)
        y = random.randint(0, img_orig.shape[0] - patch_size)
        img = img_orig[y:y+patch_size, x:x+patch_size].transpose(2,0,1)
 
        X_train.append(img)
    
    X_train = np.array(X_train)
    Y_train = X_train.copy()
        
    return X_train,Y_train




def test_image_loader(path="/home/ml/Akin/Kodak"):
    images = natsorted(glob.glob(path+"/*.png"))
        
    return images



def prepare_test_data(im):
    X_test = []
    
    img = imageio.imread(im).transpose(2,0,1)
    
    X_test.append(img)
    
    X_test = np.array(X_test)
    Y_test = X_test.copy()
    
    return X_test,Y_test



def train_one_step(im_batch,model,optimizer,criterion,aux_optimizer):

    model = model.train()
    
    X_train,Y_train = prepare_train_data(im_batch)
    X_train = normalize(torch.from_numpy(X_train).to(device).float())
    Y_train = normalize(torch.from_numpy(Y_train).to(device).float())
    
    output = model.forward(X_train) 
    
    loss_dict = criterion(output, Y_train)
    
    loss = loss_dict["loss"]
    dist_loss = loss_dict["mse_loss"]
    rate_loss = loss_dict["bpp_loss"]
    
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
    
    aux_optimizer.zero_grad()
    aux_loss = model.aux_loss()
    aux_loss.backward()
    aux_optimizer.step()
    
    
    return dist_loss.item(),rate_loss.item(),loss.item(),aux_loss.item()



def validate(model,criterion):

    with torch.no_grad():
        
        average_distortion = 0
        
        average_mse = 0
        
        total_pixels = 0

        average_test_loss = 0
        average_bpp = 0
        
        
        model = model.eval()
        
        test_images = test_image_loader()

        total_images = len(test_images)
        
        d = 0
        os.makedirs("Created/",exist_ok=True)

        for ims in test_images:
            X_test,Y_test = prepare_test_data(ims)
            
            (m,c,w,h) = X_test.shape
            total_pixels += m*w*h
            
            X_test_ = normalize(torch.from_numpy(X_test).to(device).float())
            Y_test_ = normalize(torch.from_numpy(Y_test).to(device).float())
                    
            output = model.forward(X_test_)
            out = output["x_hat"]
            
            loss_dict = criterion(output, Y_test_)
            loss = loss_dict["loss"]
            dist_loss = loss_dict["mse_loss"]
            rate_loss = loss_dict["bpp_loss"]
            
                
            average_test_loss += loss.item()
            average_bpp += rate_loss.item()
            
            uint8_out = float_to_uint8(out[0,:].cpu().detach().numpy())
            uint8_real = float_to_uint8(Y_test_[0,:].cpu().numpy())
            
            o_f = uint8_out.astype(np.float32)/255.
            r_f = uint8_real.astype(np.float32)/255.
                        
            mse_dec = MSE(r_f,o_f)
                
            average_mse += mse_dec
                                
            save_path = "Created/"+"img"+str(d)+".png"
            
            imageio.imsave(save_path,uint8_out)
                                
            d += 1
                
    average_mse /= total_images
    
    average_psnr = PSNR(average_mse,data_range=1)
    
    average_bpp /= total_images
    
    average_test_loss /= total_images
        
 
    return average_test_loss,average_psnr,average_bpp

    
    
def save_model(model,optimizer,name):
    state = {
        "state_dict" : model.state_dict(),
        "optimizer" : optimizer.state_dict()
    }
    torch.save(state,name+".pth")
    


def main():   
    np.random.seed(1)
    torch.manual_seed(2)
    random.seed(3)
    

    total_train_step = 2000000
    train_step = 1000
    lr_step = 1000000
    learning_rate = 5.e-5
    model = MeanScaleHyperprior(192,192)
    model = model.to(device).float()
    
    criterion = RateDistortionLoss(lmbda=0.0018).to(device).float()
    
    
    optimizer = torch.optim.Adam(filter(lambda p: p.requires_grad, model.parameters()), lr=learning_rate)
    aux_optimizer = torch.optim.Adam(model.aux_parameters(), lr=5.e-5)

    
    model_parameters = filter(lambda p: p.requires_grad, model.parameters())
    params = sum([np.prod(p.size()) for p in model_parameters])
    logging.info("number of compress parameters: "+str(params))
    

    
    train_images = train_image_loader()
        

    batch_size = 16
    
    
    step_train_loss = 0
    step_distortion_loss = 0
    step_rate_loss = 0
    step_aux_loss = 0
    
    best_test_loss = 10**10
    
    time_start = time.time()
    
    for minibatch_processed in range(1,total_train_step+1):
        
        image_batch = random.sample(train_images, batch_size)
        
        total_distortion_loss,total_rate_loss,loss,aux_loss = train_one_step(image_batch,model,optimizer,criterion,aux_optimizer)
        
        step_train_loss += loss
        step_distortion_loss += total_distortion_loss
        step_rate_loss += total_rate_loss 
        step_aux_loss += aux_loss
                
    
        if minibatch_processed % train_step == 0:
            average_test_loss,average_psnr_dec,average_bpp = validate(model,criterion)
            if average_test_loss < best_test_loss:
                save_model(model,optimizer,"compression")
                best_test_loss = average_test_loss
                psnr_dec_at_best_loss = average_psnr_dec
                bpp_at_best_loss = average_bpp
                logging.info("NEW BEST !!!")
            
            logging.info("learning rate: "+str(learning_rate))
            logging.info("iterations: "+str(minibatch_processed))
            logging.info("distortion loss: "+str(step_distortion_loss/train_step))
            logging.info("rate loss: "+str(step_rate_loss/train_step))
            logging.info("train loss: "+str(step_train_loss/train_step))
            logging.info("aux loss: "+str(step_aux_loss/train_step))
            logging.info("--------")
            logging.info("test psnr dec: "+str(average_psnr_dec))
            logging.info("test bpp: "+str(average_bpp))
            logging.info("test loss: "+str(average_test_loss))
            logging.info("best test loss: "+str(best_test_loss))
            logging.info("^^^")
            logging.info("psnr dec at best test loss: "+str(psnr_dec_at_best_loss))
            logging.info("bpp at best test loss: "+str(bpp_at_best_loss))
            logging.info("**************")
            
            step_train_loss = 0
            step_distortion_loss = 0
            step_rate_loss = 0
            step_aux_loss = 0
            
        if minibatch_processed % lr_step == 0:
            learning_rate = 1.e-5
            for g in optimizer.param_groups:
                g["lr"] = learning_rate
            

            
        
        
if __name__=='__main__':
    main()
```

I first set learning rate to 1e-4 but my training loss blows up somewhere. So I set it to 5e-5.

**Comment by jbegaint** [comment]
to solve the stability issues you can use gradient clippping on the norm (like here https://github.com/InterDigitalInc/CompressAI/blob/master/examples/train.py#L123). We've had good results with `torch.nn.utils.clip_grad_norm_(model.parameters(), 1)`
I was going to suggest to set the learning rate to 1e-5 after enough iterations, but looks like your code is already doing this.
if you are training the models from scratch I would recommend using a stronger learning rate for the auxiliary optimizer (like 1e-3)

**Comment by mustafaakinyilmaz** [comment]
Thank you Jean. One last question. Do you apply downsampling to vimeo90k set to reduce compression artefacts? Is it also enough to use random cropping only or do you apply flipping, rotation etc ?

**Comment by jbegaint** [comment]
We simply extract random 256x256 cropped patches from the vimeo90k triplet dataset (http://toflow.csail.mit.edu/index.html#triplet). The triplet frames were already resized to 448x256 by the dataset authors. We don't apply any flipping or rotation.

**Comment by mustafaakinyilmaz** [comment]
Thank you very much for your quick reply and this amazing work again.

**Comment by jbegaint** [comment]
Thanks! I'm closing this for now. Feel free to re-open an issue if you still encounter issues during training.
